lib/hanami/slice.rb

```diff
@@ -111,47 +111,69 @@ def define_container
+              # classes inside the slice's namespace.
+              #
+              # e.g. "lib/foo.rb" should define SliceNamespace::Foo, to be registered as
+              # "foo"
+              component_dir.namespaces.delete_root
+              component_dir.namespaces.add_root(key: nil, const: namespace_path)
+
+              config.component_dirs.add(component_dir)
 
-            config.component_dirs.add(slice_dir) do |component_dir|
-              # Expect component files in the root of these component dirs to define
-              # classes inside a namespace matching the dir.
+              application.autoloader.push_dir(root.join("lib"), namespace: namespace)
+            else
+              # Expect component files in the root of non-lib/ component dirs to define
+              # classes inside a namespace matching that dir.
               #
-              # e.g. "actions/foo.rb" should define SliceNamespace::Actions::Foo, and
-              # will be registered as "actions.foo"
+              # e.g. "actions/foo.rb" should define SliceNamespace::Actions::Foo, to be
+              # registered as "actions.foo"
 
-              dir_namespace_path = File.join(namespace_path, slice_dir)
+              dir_namespace_path = File.join(namespace_path, component_dir.path)
 
               autoloader_namespace = begin
                 inflector.constantize(inflector.camelize(dir_namespace_path))
               rescue NameError
-                namespace.const_set(inflector.camelize(slice_dir), Module.new)
+                namespace.const_set(inflector.camelize(component_dir.path), Module.new)
               end
 
-              component_dir.namespaces.root(const: dir_namespace_path, key: slice_dir)
+              component_dir.namespaces.delete_root
+              component_dir.namespaces.add_root(const: dir_namespace_path, key: component_dir.path) # TODO: do we need to swap path delimiters for key delimiters here?
+
+              config.component_dirs.add(component_dir)
 
               application.autoloader.push_dir(
-                container.root.join(slice_dir),
+                container.root.join(component_dir.path),
                 namespace: autoloader_namespace
               )
             end
           end
+
+          # Pass configured autoload dirs to the autoloader
+          application.configuration.source_dirs.autoload_paths.each do |autoload_path|
+            next unless root.join(autoload_path).directory?
+
+            dir_namespace_path = File.join(namespace_path, autoload_path)
+
+            autoloader_namespace = begin
+              inflector.constantize(inflector.camelize(dir_namespace_path))
+            rescue NameError
+              namespace.const_set(inflector.camelize(autoload_path), Module.new)
+            end
+
+            application.autoloader.push_dir(
+              container.root.join(autoload_path),
+              namespace: autoloader_namespace
+            )
+          end
```

Comment:
The amount of setup code we have here is getting pretty out of hand, and I'd like to find ways to better structure it, but I'd like to that as a separate task in the future, so for simplicity's sake I've just added this extra stanza at the bottom of all the existing setup code.